pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,200 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+
+import com.google.common.primitives.UnsignedLong;
+import io.reactivex.disposables.Disposable;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.Deposit;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class DepositRequestManager {
+
+  private final Web3j web3j;
+  private final Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer;
+  private final DepositContract depositContract;
+  private static final Logger LOG = LogManager.getLogger();
+
+  private boolean active = false;
+  private boolean requestQueued = false;
+  private BigInteger latestCanonicalBlockNumber;
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+  private Disposable newBlockSubscription;
+
+  public DepositRequestManager(
+      Web3j web3j,
+      Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer,
+      DepositContract depositContract) {
+    this.web3j = web3j;
+    this.depositsFromBlockEventConsumer = depositsFromBlockEventConsumer;
+    this.depositContract = depositContract;
+  }
+
+  public void start() {
+    newBlockSubscription =
+        web3j
+            .blockFlowable(false)
+            .map(EthBlock::getBlock)
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(
+                this::updateLatestCanonicalBlockNumber,
+                err -> {
+                  LOG.warn("New block subscription failed, retrying.", err);
+                  start();
+                });
+  }
+
+  public void stop() {
+    newBlockSubscription.dispose();
+  }
+
+  private synchronized void updateLatestCanonicalBlockNumber(
+      BigInteger latestCanonicalBlockNumber) {
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    if (active) {
+      requestQueued = true;
+    } else {
+      active = true;
+      getLatestDeposits();
+    }
+  }
+
+  private void getLatestDeposits() {
+    requestQueued = false;
```

Comment:
This needs to be in a `synchronized (this)` block.

Reply:
`getLatestDeposits` is only called inside two synchronized blocks, doesn't that make sure we this whole method is executed by only one thread at a time anyway?

Reply:
actually nevermind, its also called from `onDepositRequestFailed`, which is not `synchronized`, is that the reason we need the `synchronized(this)` block here? and also why not use the `volatile` keyword instead?

Reply:
Done, but would appreciate your explanation

Reply:
We can't use `volatile` because `active` and `requestQueued` need to be changed in-sync - particularly the `if (active) requestQueued = true` on lines 84-85 need to be an atomic block but can't be if a different thread changes `requestQueued` without acquiring the same lock via synchronised.

If `getLatestDeposits` was only called from within synchronised blocks then this wouldn't be required.